vm/src/cformat.rs

```diff
@@ -0,0 +1,357 @@
+        }
+    }
+
+    fn parse_specifier(text: &str) -> Result<(CFormatPart, &str, usize), ParsingError> {
+        let consumed = 0;
+        let mut chars = text.chars();
+        if chars.next() != Some('%') {
+            return Err((CFormatErrorType::MissingModuloSign, consumed));
+        }
+        let consumed = consumed + 1;
+
+        let (mapping_key, after_mapping_key) =
+            CFormatString::parse_spec_mapping_key(chars.as_str()).map_err(|err| (err, consumed))?;
+        let consumed = text.find(after_mapping_key).unwrap();
+        let (flags, after_flags) = CFormatString::parse_flags(after_mapping_key);
+        let (width, after_width) = parse_number(after_flags);
+        let (precision, after_precision) = parse_precision(after_width);
+        // A length modifier (h, l, or L) may be present,
+        // but is ignored as it is not necessary for Python – so e.g. %ld is identical to %d.
+        let after_length = CFormatString::consume_length(after_precision);
+        let (format_type, remaining_text, format_char) =
+            CFormatString::parse_format_type(after_length).map_err(|err| (err, consumed))?;
+        let consumed = text.find(remaining_text).unwrap();
+
+        // apply default precision for float types
+        let precision = match precision {
+            Some(precision) => Some(precision),
+            None => match format_type {
+                CFormatType::Float(_) => Some(6),
+                _ => None,
+            },
+        };
+
+        Ok((
+            CFormatPart::Spec(CFormatSpec {
+                mapping_key: mapping_key,
+                flags: flags,
+                min_field_width: width,
+                precision: precision,
+                format_type: format_type,
+                format_char: format_char,
+            }),
+            remaining_text,
+            consumed,
+        ))
+    }
+}
+
+impl FromStr for CFormatString {
+    type Err = CFormatError;
+
+    fn from_str(text: &str) -> Result<Self, Self::Err> {
+        let mut cur_text: &str = text;
+        let mut index = 0;
+        let mut parts: Vec<(usize, CFormatPart)> = Vec::new();
+        while !cur_text.is_empty() {
+            cur_text = CFormatString::parse_literal(cur_text)
+                .or_else(|_| CFormatString::parse_specifier(cur_text))
+                .map(|(format_part, new_text, consumed)| {
+                    parts.push((index, format_part));
+                    index = index + consumed;
+                    new_text
+                })
+                .map_err(|(e, consumed)| {
+                    println!("parts {:?}", parts);
+                    CFormatError {
+                        typ: e,
+                        index: index + consumed,
+                    }
+                })?;
+        }
+
+        Ok(CFormatString {
+            format_parts: parts,
+        })
+    }
+}
```

Comment:
I propose to add some (perhaps 3 or 7) test cases in this file as unittests.

Reply:
done :)